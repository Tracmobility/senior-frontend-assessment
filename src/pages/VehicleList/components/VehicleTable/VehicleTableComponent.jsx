import React from 'react';
import PropTypes from 'prop-types';
import Loader from 'react-loader-spinner';
import { MDBDataTable } from 'mdbreact';
import { wordFormatter, batteryFormatter } from '../../../../utils/tableUtils';
import "react-loader-spinner/dist/loader/css/react-spinner-loader.css";
import styles from './VehicleTableComponent.module.scss';

const VehicleTableComponent = ({ vehicles }) => {
  if (vehicles.content) {
    const rowData = vehicles.content.map(content => {
      let dataTable = {};
      dataTable = {
        vehicleID: content.uuid,
        qrCode: content.qrCode,
        status: wordFormatter(content.status),
        location: '',
        batteryLevel: batteryFormatter(content.batteryLevel),
        operation: (
          <div className={styles.operationButtons}>
            <button className={styles.opEdit} type="button">Edit</button>
            <button className={styles.opOr} type="button">Order</button>
            <button className={styles.opCity} type="button">Change City/Region</button>
            <button className={styles.opStat} type="button">Change Status</button>
          </div>
        ),
      }
      return dataTable
    });

    const data = {
      columns: [
        {
          label: 'VehicleID',
          field: 'vehicleID',
          sort: 'asc',
          width: 100,
        },
        {
          label: 'QR Code',
          field: 'qrCode',
          sort: 'asc',
          width: 300,
        },
        {
          label: 'Status',
          field: 'status',
          sort: 'asc',
          width: 150,
        },
        {
          label: 'Location',
          field: 'location',
          sort: 'asc',
          width: 250,
        },
        {
          label: 'Battery Level',
          field: 'batteryLevel',
          sort: 'asc',
          width: 120,
        },
        {
          label: 'Operation',
          field: 'operation',
          sort: 'asc',
          width: 800,
        },
      ],
      rows: rowData,
    };

    return (
      <MDBDataTable
        striped
        bordered
        small
        responsive
        data={data}
        className={styles.dataTable}
        maxHeight="200px"
      />
    );
  } else {
    return (
      <div className={styles.vehicleListLoader}>
        <Loader
          type="Grid"
          color="#03989E"
          height={100}
          width={100}
          timeout={3000}
        />
      </div>
    )
  }
};

VehicleTableComponent.propTypes = {
  vehicles: PropTypes.shape({
    content: PropTypes.arrayOf(
      PropTypes.shape({
        uuid: PropTypes.string,
        qrCode: PropTypes.string,
        status: PropTypes.string,
        batteryLevel: PropTypes.number,
      }),
    ),
  })
};

export default VehicleTableComponent;