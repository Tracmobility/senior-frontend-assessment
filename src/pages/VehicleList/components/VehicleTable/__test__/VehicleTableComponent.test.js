import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import VehicleTableComponent from '../VehicleTableComponent';

Enzyme.configure({ adapter: new Adapter() });

const vehicles = {
  content: [
    {
      "uuid": "RTyclaTkpV",
      "createdAt": "2010-02-15T00:09:01",
      "updatedAt": "2013-01-19T03:39:46",
      "frameNo": "lpjOX",
      "qrCode": "KNaWwUOWdve",
      "status": "MAINTENANCE_PENDING",
      "lockStatus": "UNLOCKED",
      "lastSyncAt": "2017-01-07T08:25:16",
      "category": "EKICKSCOOTER",
      "discount": 0.7689465141194151,
      "graphUrl": "dIrQEcHPrTBIuosqKfVJaYdtmUlp",
      "batteryLevel": 0.5512746,
      "imei": "WgNOgmbaRybbeJkhxuSiYrYz",
      "cleanedAt": "2017-04-10T08:43:44"
    }
  ]
}

const vehicleTableComponent = (<VehicleTableComponent  vehicles={vehicles} />);

describe('It matches the previous working version of the table component', () => {
  it('Matches the previous snapshot of the Vehicle Table Component', () => {
    const tree = renderer.create(vehicleTableComponent).toJSON();
    expect(tree).toMatchSnapshot();
  });
});