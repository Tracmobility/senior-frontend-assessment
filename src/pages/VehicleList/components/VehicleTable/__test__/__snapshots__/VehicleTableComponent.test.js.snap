// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It matches the previous working version of the table component Matches the previous snapshot of the Vehicle Table Component 1`] = `
<div
  className="dataTables_wrapper dt-bootstrap4 dataTable"
  data-test="datatable"
>
  <div
    className="row"
  >
    <div
      className="col-sm-12 col-md-6"
      data-test="datatable-entries"
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": false,
        }
      }
    >
      <div
        className="dataTables_length bs-select"
        data-test="datatable-select"
      >
        <label>
          Show entries
          <select
            className="custom-select custom-select-sm form-control form-control-sm"
            onChange={[Function]}
            style={
              Object {
                "marginLeft": ".5rem",
              }
            }
            value={10}
          >
            <option
              value={10}
            >
              10
            </option>
            <option
              value={20}
            >
              20
            </option>
            <option
              value={50}
            >
              50
            </option>
            <option
              value={100}
            >
              100
            </option>
          </select>
        </label>
      </div>
    </div>
    <div
      className="col-sm-12 col-md-6"
      data-test="datatable-search"
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "flex-end",
        }
      }
    >
      <div
        className="mdb-datatable-filter flex-row"
        data-test="datatable-input"
      >
        <input
          aria-label="Search"
          className="form-control form-control-sm ml-0 my-1"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-sm-12"
      data-test="datatable-table"
    >
      <div
        className="table-responsive"
        data-test="table"
        style={
          Object {
            "maxHeight": undefined,
          }
        }
      >
        <table
          className="table table-bordered table-sm table-striped dataTable"
          entries={10}
        >
          <thead
            data-test="datatable-head"
          >
            <tr>
              <th
                className="sorting"
                onClick={[Function]}
              >
                VehicleID
              </th>
              <th
                className="sorting"
                onClick={[Function]}
              >
                QR Code
              </th>
              <th
                className="sorting"
                onClick={[Function]}
              >
                Status
              </th>
              <th
                className="sorting"
                onClick={[Function]}
              >
                Location
              </th>
              <th
                className="sorting"
                onClick={[Function]}
              >
                Battery Level
              </th>
              <th
                className="sorting"
                onClick={[Function]}
              >
                Operation
              </th>
            </tr>
          </thead>
          <tbody
            data-test="table-body"
          >
            <tr>
              <td>
                RTyclaTkpV
              </td>
              <td>
                KNaWwUOWdve
              </td>
              <td>
                Maintenance Pending
              </td>
              <td>
                
              </td>
              <td>
                55%
              </td>
              <td>
                <div
                  className="operationButtons"
                >
                  <button
                    className="opEdit"
                    type="button"
                  >
                    Edit
                  </button>
                  <button
                    className="opOr"
                    type="button"
                  >
                    Order
                  </button>
                  <button
                    className="opCity"
                    type="button"
                  >
                    Change City/Region
                  </button>
                  <button
                    className="opStat"
                    type="button"
                  >
                    Change Status
                  </button>
                </div>
              </td>
            </tr>
          </tbody>
          <thead
            data-test="table-foot"
          >
            <tr>
              <th>
                VehicleID
              </th>
              <th>
                QR Code
              </th>
              <th>
                Status
              </th>
              <th>
                Location
              </th>
              <th>
                Battery Level
              </th>
              <th>
                Operation
              </th>
            </tr>
          </thead>
        </table>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-sm-12 col-md-5"
      data-test="datatable-info"
    >
      <div
        aria-live="polite"
        className="dataTables_info"
        role="status"
      >
        Showing 1 to 1 of 1 entries
      </div>
    </div>
    <div
      className="col-sm-12 col-md-7"
      data-test="datatable-pagination"
    >
      <div
        className="dataTables_paginate"
      >
        <ul
          className="pagination"
          data-test="pagination"
        >
          <li
            className="disabled page-item"
            data-test="page-item"
          >
            <a
              aria-label="Previous"
              className="page-link page-link"
              data-test="page-link"
              onClick={[Function]}
            >
              <span>
                Previous
              </span>
            </a>
          </li>
          <li
            className="active page-item"
            data-test="page-item"
          >
            <a
              className="page-link page-link"
              data-test="page-link"
              onClick={[Function]}
            >
              1
              <span
                className="sr-only"
              >
                (current)
              </span>
            </a>
          </li>
          <li
            className="disabled page-item"
            data-test="page-item"
          >
            <a
              aria-label="Next"
              className="page-link page-link"
              data-test="page-link"
              onClick={[Function]}
            >
              <span>
                Next
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
